geode-apis-compatible-with-redis/src/integrationTest/java/org/apache/geode/redis/internal/executor/pubsub/AbstractSubCommandsIntegrationTest.java

```diff
@@ -0,0 +1,203 @@
+    introspector = new Jedis("localhost", getPort());
+  }
+
+  @After
+  public void teardown() {
+    if (mockSubscriber.getSubscribedChannels() > 0) {
+      mockSubscriber.unsubscribe();
+    }
+    waitFor(() -> mockSubscriber.getSubscribedChannels() == 0);
+  }
+
+  @Test
+  public void pubsub_shouldError_givenTooFewArguments() {
+    assertAtLeastNArgs(introspector, Protocol.Command.PUBSUB, 1);
+  }
+
+  @Test
+  public void channels_shouldError_givenTooManyArguments() {
+    assertAtMostNArgsForSubCommand(introspector,
+        Protocol.Command.PUBSUB,
+        "channels".getBytes(),
+        1);
+  }
+
+  @Test
+  public void pubsub_shouldReturnError_givenUnknownSubcommand() {
+    String expected = String.format(ERROR_UNKNOWN_PUBSUB_SUBCOMMAND, "nonesuch");
+
+    assertThatThrownBy(() -> introspector.sendCommand(Protocol.Command.PUBSUB, "nonesuch"))
+        .hasMessageContaining(expected);
+  }
+
+  @Test
+  @SuppressWarnings("unchecked")
+  public void channels_shouldReturnListOfAllChannels_withActiveChannelSubscribers_whenCalledWithoutPattern() {
+    ArrayList<byte[]> expectedChannels = new ArrayList<>();
+    expectedChannels.add("foo".getBytes());
+    expectedChannels.add("bar".getBytes());
+    Runnable runnable =
+        () -> subscriber.subscribe(mockSubscriber, "foo", "bar");
+    Thread subscriberThread = new Thread(runnable);
+
+    subscriberThread.start();
+    waitFor(() -> mockSubscriber.getSubscribedChannels() == 2);
+    ArrayList<byte[]> result = (ArrayList<byte[]>) introspector
+        .sendCommand(Protocol.Command.PUBSUB, "channels");
+
+    assertThat(result).containsExactlyInAnyOrderElementsOf(expectedChannels);
+
+  }
+
+  @Test
+  public void channels_shouldNeverReturnPsubscribedChannels_givenNoActiveChannelSubscribers() {
+
+    Runnable runnable = () -> subscriber.psubscribe(mockSubscriber, "f*");
+    Thread patternSubscriberThread = new Thread(runnable);
+
+    patternSubscriberThread.start();
+    waitFor(() -> mockSubscriber.getSubscribedChannels() == 1);
+    ArrayList<String> result =
+        (ArrayList<String>) introspector.pubsubChannels("f*");
+
+    assertThat(result).isEmpty();
+
+    mockSubscriber.punsubscribe();
+  }
+
+
+  @Test
+  public void channels_shouldReturnListOfMatchingChannels_withActiveChannelSubscribers_whenCalledWithPattern() {
+
+    ArrayList<String> expectedChannels = new ArrayList<>();
+    expectedChannels.add("foo");
+
+    Runnable runnable =
+        () -> subscriber.subscribe(mockSubscriber, "foo", "bar");
+
+    Thread subscriberThread = new Thread(runnable);
+    subscriberThread.start();
+    waitFor(() -> mockSubscriber.getSubscribedChannels() == 2);
+    ArrayList<String> result =
+        (ArrayList<String>) introspector.pubsubChannels("fo*");
+
+    assertThat(result).containsExactlyInAnyOrderElementsOf(expectedChannels);
+  }
+
+  @Test
+  public void channels_should_returnEmptyArry_GivenPatternWithNoMatches() {
```

Comment:
Couple of typos here, should be "returnEmptyArray" and for consistency, "given" with a lower-case g.